confmap/expand.go

```diff
@@ -0,0 +1,213 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package confmap // import "go.opentelemetry.io/collector/confmap"
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+func (mr *Resolver) expandValueRecursively(ctx context.Context, value any) (any, error) {
+	for i := 0; i < 100; i++ {
+		val, changed, err := mr.expandValue(ctx, value)
+		if err != nil {
+			return nil, err
+		}
+		if !changed {
+			return val, nil
+		}
+		value = val
+	}
+	return nil, errTooManyRecursiveExpansions
+}
+
+func (mr *Resolver) expandValue(ctx context.Context, value any) (any, bool, error) {
+	switch v := value.(type) {
+	case string:
+		if !strings.Contains(v, "${") || !strings.Contains(v, "}") {
+			// No URIs to expand.
+			return value, false, nil
+		}
+
+		if strings.Count(v, "}") > 100 {
+			// Too many closing brackets. Don't expand to protect from too deep recursion.
+			return value, false, nil
+		}
+
+		URI := findURI(v)
+		if URI != "" && URI == value {
+			// If the value is a single URI, then the return value can be anything.
+			// This is the case `foo: ${file:some_extra_config.yml}`.
+			return mr.expandStringURI(ctx, v)
+		}
+
+		// Embedded or nested URIs.
+		return mr.findAndExpandURI(ctx, v)
+	case []any:
+		nslice := make([]any, 0, len(v))
+		nchanged := false
+		for _, vint := range v {
+			val, changed, err := mr.expandValue(ctx, vint)
+			if err != nil {
+				return nil, false, err
+			}
+			nslice = append(nslice, val)
+			nchanged = nchanged || changed
+		}
+		return nslice, nchanged, nil
+	case map[string]any:
+		nmap := map[string]any{}
+		nchanged := false
+		for mk, mv := range v {
+			val, changed, err := mr.expandValue(ctx, mv)
+			if err != nil {
+				return nil, false, err
+			}
+			nmap[mk] = val
+			nchanged = nchanged || changed
+		}
+		return nmap, nchanged, nil
+	}
+	return value, false, nil
+}
+
+// findURI finds the URI corresponding to the first closing bracket in input. It returns
+// the URI if it is expandable, or an empty string if it is not expandable.
+// Note: findURI is only called when input contains a closing bracket.
+func findURI(input string) string {
+	closeIndex := closeIndex(input)
+	openIndex := openIndex(input[:closeIndex+1])
+	if openIndex < 0 {
+		// Should not expand because there is a missing ${.
+		return ""
+	}
+
+	URI := input[openIndex : closeIndex+1]
+
+	if !strings.Contains(URI, ":") {
+		// Should not expand. This is expanded in the expandconverter.
+		return ""
+	}
```

Comment:
> That checks for the uri structure.

Do you mean by calling function `newLocation` which uses the `uriRegexp` to check the URI structure ?

I don't believe we should check the uri structure through `uriRegexp` for this case ? The `!strings.Contains(URI, ":")` is to check if this is the old way to expand env vars (e.g. `${HOST}`). If I pass this over to function `newLocation`, then it will throw an invalid uri error, although this syntax is currently expected.

Am I misunderstanding the suggestion ?